rust/noosphere/src/ffi/fs.rs

```diff
@@ -0,0 +1,233 @@
+        .unwrap()
+}
+
+#[ffi_export]
+/// De-allocate an [NsSphereFs] instance
+pub fn ns_sphere_fs_free(sphere_fs: repr_c::Box<NsSphereFs>) {
+    drop(sphere_fs)
+}
+
+#[ffi_export]
+/// Read a [NsSphereFile] from a [NsSphereFs] instance by slashlink. Note that
+/// although this function will eventually support slashlinks that include the
+/// pet name of a peer, at this time only slashlinks with slugs referencing the
+/// slug namespace of the local sphere are allowed.
+pub fn ns_sphere_fs_read(
+    noosphere: &NsNoosphereContext,
+    sphere_fs: &NsSphereFs,
+    slashlink: char_p::Ref<'_>,
+) -> Option<repr_c::Box<NsSphereFile>> {
+    noosphere
+        .async_runtime()
+        .block_on(async {
+            let slashlink = Slashlink::from_str(slashlink.to_str())?;
+
+            if Peer::None != slashlink.peer {
+                return Err(anyhow!("Peer in slashlink not yet supported"));
+            }
+
+            let slug = match slashlink.slug {
+                Some(slug) => slug,
+                None => return Err(anyhow!("No slug specified in slashlink!")),
+            };
+
+            println!(
+                "Reading sphere {} slug {}...",
+                sphere_fs.inner().identity(),
+                slug
+            );
+
+            let file = sphere_fs.inner().read(&slug).await?;
+
+            Ok(file.map(|sphere_file| {
+                repr_c::Box::new(NsSphereFile {
+                    inner: sphere_file.boxed(),
+                })
+            }))
+        })
+        .unwrap()
+}
+
+#[ffi_export]
+/// Write a byte buffer to a slug in the given [NsSphereFs] instance, assigning
+/// its content-type header to the specified value. If additional headers are
+/// specified, they will be appended to the list of headers in the memo that is
+/// created for the content. If some content already exists at the specified
+/// slug, it will be assigned to be the previous historical revision of the new
+/// content.
+///
+/// Note that you must invoke [ns_sphere_fs_save] to commit one or more writes
+/// to the sphere.
+pub fn ns_sphere_fs_write(
+    noosphere: &NsNoosphereContext,
+    sphere_fs: &mut NsSphereFs,
+    slug: char_p::Ref<'_>,
+    content_type: char_p::Ref<'_>,
+    bytes: c_slice::Ref<'_, u8>,
+    additional_headers: Option<&NsHeaders>,
+) {
+    noosphere.async_runtime().block_on(async {
+        let slug = slug.to_str();
+
+        println!(
+            "Writing sphere {} slug {}...",
+            sphere_fs.inner().identity(),
+            slug
+        );
+
+        match sphere_fs
+            .inner_mut()
+            .write(
+                slug,
+                content_type.to_str().try_into().unwrap(),
+                bytes.as_ref(),
+                additional_headers.map(|headers| headers.inner().clone()),
+            )
+            .await
+        {
+            Ok(_) => println!("Updated {:?}...", slug),
+            Err(error) => println!("Sphere write failed: {}", error),
```

Comment:
Maybe more discussion needed here, but how do we generally translate fallible Rust calls over FFI? For this method, we could return the number of written bytes so there's a programmatic way of observing runtime failures -- e.g. maybe this method isn't actually writing to disk, where a "reasonable failure" could occur, but could be represented as a `bool` return value. I think it's common for C libs to have at the very least booleans on every fallible method (or an int enum with `NsOK` or `NsFailure` or something) due to lack of catching. Creation methods can return `NULL` on failed calls indicating failure (documented as such).

Not suggesting this at the moment but illustrating possibilities, to have a more robust error mechanism, e.g. `NsResponse` on every method call with `NsFailure => 0, NsOk => 1, NsFsWriteError = 2` and so forth, we'd also need method calls taking addresses to allocate into, which would be a larger change:
```c
NsSphereFs* sphere_fs;
if (NsOk != ns_sphere_fs_open(&sphere_fs)) {
  // handle error, `sphere_fs` is NULL
}
```

Reply:
Great question. I don't have a good answer for this, and have kicked this can down the road. Suggestions welcome!

https://github.com/subconsciousnetwork/noosphere/issues/145